### PR TITLE
[SYCL] Refactor OpenCL kernel generation

### PR DESCRIPTION
**Pull request by Fznamznon**

### CODE REVIEW

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -376,67 +374,68 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
...
+  std::pair<DeclaratorDecl *, DeclaratorDecl *> MappingPair;
   Sema &SemaRef;
 };
 
-static FunctionDecl *CreateSYCLKernelFunction(ASTContext &Context,
-                                              StringRef Name,
-                                              ArrayRef<ParamDesc> ParamDescs) {
+static FunctionDecl *
+CreateOpenCLKernelDeclaration(ASTContext &Context, StringRef Name,
+                            ArrayRef<ParamDesc> ParamDescs) {
```
Comment (agozillon) [inline clang/lib/Sema/SemaSYCL.cpp]: Indentation is a bit off I think

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -376,67 +374,68 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
...
-  DeclMap DMap;
+  std::pair<DeclaratorDecl *, DeclaratorDecl *> MappingPair;
   Sema &SemaRef;
 };
 
-static FunctionDecl *CreateSYCLKernelFunction(ASTContext &Context,
-                                              StringRef Name,
-                                              ArrayRef<ParamDesc> ParamDescs) {
+static FunctionDecl *
+CreateOpenCLKernelDeclaration(ASTContext &Context, StringRef Name,
```
Comment (agozillon) [inline clang/lib/Sema/SemaSYCL.cpp]: Maybe it's worth commenting this function similarly to CreateOpenCLKernelBody while we are here?

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Hm, In comment for `CreateOpenCLKernelBody ` I described what we do inside the OpenCL kernel body but I don't see what I can write for `CreateOpenCLKernelDeclaration`. Do you have any suggestions?

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Yes, I will fix it, thanks.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 2 above)
Comment (agozillon) [reply clang/lib/Sema/SemaSYCL.cpp]: Sorry, In hindsight I think your correct the function name/definition itself is pretty self explanatory, so you can disregard my comment, apologies for wasting your time in this case.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Done.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -847,18 +778,18 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
...
-      const auto *AccTy = ArgTy->getAsCXXRecordDecl();
-      assert(AccTy && "accessor must be of a record type");
-      const auto *AccTmplTy = cast<ClassTemplateSpecializationDecl>(AccTy);
-      int Dims = static_cast<int>(
-          AccTmplTy->getTemplateArgs()[1].getAsIntegral().getExtValue());
-      int Info = getAccessTarget(AccTmplTy) | (Dims << 11);
-      H.addParamDesc(SYCLIntegrationHeader::kind_accessor, Info, Offset);
+    // The parameter is a SYCL accessor object.
+    // The Info field of the parameter descriptor for accessor contains
+    // two template parameters packed into thid integer field:
```
Comment (keryell) [inline clang/lib/Sema/SemaSYCL.cpp]: thid?

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -535,16 +532,13 @@ CreateSYCLKernelBody(Sema &S, FunctionDecl *KernelCallerFunc, DeclContext *DC) {
...
-        ParamStmts.push_back(getExprForRangeOrOffset(
-            S, ((*ParamItr++))->getOriginalType(), ParamDREs[2]));
-        ParamStmts.push_back(getExprForRangeOrOffset(
-            S, ((*ParamItr++))->getOriginalType(), ParamDREs[3]));
+        llvm::SmallVector<Expr *, 4> ParamStmts;
+        for (size_t I = 0; I < NumParams; ++I) {
+          ParamStmts.push_back(getExprForKernelParameter(
+              S, (*(ParamItr++))->getOriginalType(), ParamDREs[I]));
+        }
         // [kenrel_obj or wrapper object].accessor.__init(_ValueType*,
```
Comment (keryell) [inline clang/lib/Sema/SemaSYCL.cpp]: Since you are visiting the area :-) what about `kernel_obj`?

File: clang/lib/Sema/SemaSYCL.cpp [hunk 5]
```diff
@@ -449,79 +448,77 @@ static CXXMethodDecl *getInitMethod(const CXXRecordDecl *CRD) {
...
         }
-        std::advance(TargetFuncParam, NumParams - 1);
+        std::advance(KernelFuncParam, NumParams - 1);
 
         DeclAccessPair FieldDAP = DeclAccessPair::make(Field, AS_none);
-        // [kenrel_obj or wrapper object].accessor
-        auto AccessorME = MemberExpr::Create(
-            S.Context, Base, false, SourceLocation(),
-            NestedNameSpecifierLoc(), SourceLocation(), Field, FieldDAP,
+        // [kenrel_obj or wrapper object].special_obj
```
Comment (bader) [inline clang/lib/Sema/SemaSYCL.cpp]:
```suggestion
        // [kernel_obj or wrapper object].special_obj
```

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (bader) [inline clang/lib/Sema/SemaSYCL.cpp]:
```suggestion
        // [kernel_obj or wrapper object].accessor.__init(_ValueType*,
```

File: clang/lib/Sema/SemaSYCL.cpp (hunk 3 above)
Comment (bader) [inline clang/lib/Sema/SemaSYCL.cpp]:
```suggestion
    // two template parameters packed into an integer field:
```

File: clang/lib/Sema/SemaSYCL.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Thanks for nice catch! Done.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 3 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Fixed.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: No problem. Done.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 5 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Done.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Done.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 3 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Done.

**Pull request merged**